core/DCache.cpp

```diff
@@ -87,37 +87,65 @@ namespace olympia {
         return cache_hit;
     }
 
-    const uint32_t DCache::mshrLookup_(const MemoryAccessInfoPtr &mem_access_info_ptr) {
-        // TODO
-        return 0;
+    sparta::Buffer<DCache::MSHREntryInfoPtr>::iterator DCache::mshrLookup_(const uint64_t& block_address) {
+        sparta::Buffer<MSHREntryInfoPtr>::iterator it;
+
+        for(it = mshr_file_.begin(); it < mshr_file_.end(); it++) {
+            auto mshr_block_addr = (*it)->getBlockAddress();
+            if(mshr_block_addr == block_address){
+                // Address match
+                break;
+            }
+        }
+
+        return it;
     }
 
-    const uint32_t DCache::allocateMSHREntry_(uint64_t addr){
-        // TODO
-        auto block_address = addr_decoder_->calcBlockAddr(addr);
-        // Find free entry in MSHRFile and allocate it
-        return 0;
+    sparta::Buffer<DCache::MSHREntryInfoPtr>::iterator DCache::allocateMSHREntry_(uint64_t block_address){
+
+        MSHREntryInfoPtr mshr_entry = sparta::allocate_sparta_shared_pointer<MSHREntryInfo>(mshr_entry_allocator,
+                                                                                            block_address, cache_line_size_);
+        auto it = mshr_file_.push_back(mshr_entry);
+        return it;
     }
     
     // Handle request from the LSU
-    void DCache::getInstsFromLSU_(const MemoryAccessInfoPtr &memory_access_info_ptr){
+    void DCache::processInstsFromLSU_(const MemoryAccessInfoPtr &memory_access_info_ptr){
         // Check if there is an incoming cache refill request from BIU
         if(incoming_cache_refill_) {
             // Cache refill is given priority
-            // nack signal is sent to the LSU 
+            // nack signal is sent to the LSU
             // LSU will have to replay the instruction
-            out_lsu_lookup_nack_.send(0);
+            out_lsu_lookup_nack_.send(1);
         }
         else {
             // Append to Cache Pipeline
-            cache_pipeline_.append(memory_access_info_ptr);
+            cache_pipeline_.append(memory_access_info_ptr); // Data read stage executes in the next cycle
         }
     }
 
     // Incoming Cache Refill request from BIU
-    void DCache::getAckFromBIU_(const InstPtr &inst_ptr) {
+    // Precedence: Before S1 (Cache Lookup)
+    void DCache::getRefillFromBIU_(const InstPtr &inst_ptr) {
+        // change inst_ptr to cache_line TODO
```

Comment:
`inst_ptr` is unused.  Consider changing `incoming_cache_refill_` from a bool to the instruction (set it to `nullptr` if none) being serviced for the refill and/or use the instruction pointer in an assert to make sure the `current_refill_mshr_entry_` matches the instruction.

Reply:
Understood, I will make the changes